[Issue 381] Feedback mechanism

cesarosum opened the pull request:
Adds a feedback button which will open a text box and copy the text to a file in a shared folder called "Feedback".

Please test with the corresponding branch in web-ui.

cesarosum commented:
@ianopolous Test is written but failing. I suspect the problem is on L500 where I'm getting the path to the feedback directory but not the feedback file itself. Is getDescendentByPath appropriate here?

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -490,6 +490,24 @@ public void followPeergos() throws Exception {
+            .get();
+
+        byte[] fileContents = Serialize.readFully(inputStream, feedbackFile.get().getFileProperties().size).get();
+        assertTrue("Feedback file contents correct", feedback.getBytes() == fileContents);
```
need to use Arrays.equals otherwise you're comparing object identity

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -490,6 +490,24 @@ public void followPeergos() throws Exception {
+
+        // Can peergos read the feedback file?
+        Path filePath = Paths.get(u1.username, "feedback");
+        Optional<FileWrapper> feedbackFile = u1.getByPath(filePath).get();
```
this is the feedback dir, not the file. You need to call getChildren on this to get the file

cesarosum replied:
Got it. Thanks 👍

cesarosum commented:
Ah wait, I think I have a fix.

cesarosum commented:
@ianopolous The test now runs correctly.

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -488,10 +490,43 @@ public void followPeergos() throws Exception {
         UserContext u1 = PeergosNetworkUtils.ensureSignedUp("peergos", "testpassword", network, crypto);
         UserContext u2 = PeergosNetworkUtils.ensureSignedUp("w", "w", network, crypto);
 
+        u2.sendFollowRequest(u1.username, SymmetricKey.random()).get();
```
this shouldn't be here. Users automatically friend "peergos" during signup

cesarosum replied:
Got it. Will remove.

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -488,10 +490,43 @@ public void followPeergos() throws Exception {
 
+    @Test
+    public void sendFeedbackToPeergos() throws Exception {
+        UserContext u1 = PeergosNetworkUtils.ensureSignedUp("peergos", "testpassword", network, crypto);
```
peergos is probably a clearer name than u1?

cesarosum replied:
Sure is. How about I rename `u1` to `peergos` and `u2` to `newUser`?

ianopolous commented:
The new test is failing: https://teamcity.jetbrains.com/viewLog.html?buildId=2107680&tab=buildResultsDiv&buildTypeId=OpenSourceProjects_Peergos_Build

cesarosum commented:
@ianopolous That's weird. OK I've rebased and pushed again. Test runs fine locally. For reference, I added the following to `build.xml`:
```
<target name="MultiUserTests" depends="compile,dist" description="Run the multiuser test suite.">
    <junit fork="true" printsummary="true" haltonfailure="yes">
      <jvmarg value="-Xmx2g"/>
        <classpath>
           <pathelement location="lib/jnr-fuse-0.3-all.jar" />
           <pathelement location="lib/sqlite-jdbc-3.7.2.jar" />
           <pathelement location="lib-build/junit-4.11.jar" />
           <pathelement location="lib-build/hamcrest-core-1.3.jar" />
           <pathelement location="PeergosServer.jar" />
        </classpath>
        <formatter type="plain" usefile="false"/>
        <test name="peergos.server.tests.MultiUserTests" methods="sendFeedbackToPeergos">
        </test>
   </junit>
</target>
```
and it ran successfully -- 
```
    [junit] Testcase: sendFeedbackToPeergos took 20.798 sec
    [junit] Writing 746 bytes to /var/folders/8c/mktllgc1157fnz5cb1pwryj00000gn/T/peergos158635005344135473/usage-state.cbor
    [junit] Successfully stored usage-state to /var/folders/8c/mktllgc1157fnz5cb1pwryj00000gn/T/peergos158635005344135473/usage-state.cbor

BUILD SUCCESSFUL
Total time: 1 minute 39 seconds
```

cesarosum commented:
@ianopolous It's still failing. I'll re-run the whole test suite locally to see if I can replicate. It seems to be L513 is returning empty (i.e. the shared feedback directory is empty).

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -493,6 +495,37 @@ public void followPeergos() throws Exception {
+    @Test
+    public void sendFeedbackToPeergos() throws Exception {
+        UserContext peergos = PeergosNetworkUtils.ensureSignedUp("peergos", "testpassword", network, crypto);
+        UserContext newUser = PeergosNetworkUtils.ensureSignedUp("w", "w", network, crypto);
```
best to choose a random username (random() will do) to ensure independence from other tests (we set up one cluster for the whole test suite)

ianopolous replied:
That might be why it passes when run on it's own.

cesarosum replied:
Ah! OK cool. I'll make that change now.

cesarosum replied:
Hmmm, still failing. I've replicated the error locally so will work on that later tonight.

ianopolous commented on src/peergos/server/tests/MultiUserTests.java:
```diff
@@ -493,6 +495,37 @@ public void followPeergos() throws Exception {
+        assertTrue("Feedback submission was successful!", testFeedbackSubmission.get() == true);
+
+        List<FollowRequestWithCipherText> peergosRequests = peergos.processFollowRequests().get();
+        peergos.sendReplyFollowRequest(peergosRequests.get(0), true, true).get();
```
after this you could call newUser.processFollowRequests() which finalises the bidirectional following, though it shouldn't be needed.

cesarosum commented:
@ianopolous The build has made it through the MultiUserTest suite, the only failing test is that Fuse one. I think we can consider this done now. Please merge at your earliest convenience.
